### PR TITLE
git-diff: Fix subfolder git detection

### LINKED ISSUE #1149: Does not work for Github subfolder within project directory
I develop WordPress plugins and as such my porject directory is then entire WordPress root folder, which means that my github folder is acatually a sub-folder within the wp-content/plugins/ folder.

Is there anyway to get githb.atom.io package to work from a given sub-folder?

### LINKED ISSUE #1835: Git integration doesn't work if .git/ not in top level
*@alexwhitman commented on May 13, 2014, 8:40 AM UTC:*

If the `.git/` directory isn't in the top level directory then git integration doesn't work:

[atom-git](https://cloud.githubusercontent.com/assets/422013/2954963/0ade5726-da7a-11e3-9917-63747fe240ae.png)

*This issue was moved by [rsese](https://github.com/rsese) from [atom/atom#2203](https://github.com/atom/atom/issues/2203).*

### LINKED ISSUE #21630: git-diff: Doesn't detect subfolder as repos
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: <https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode>
    * Followed all applicable steps in the debugging guide: <https://flight-manual.atom.io/hacking-atom/sections/debugging/>
    * Checked the FAQs on the message board for common solutions: <https://discuss.atom.io/c/faq>
    * Checked that your issue isn't already filed: <https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom>
    * Checked that there is not already an Atom package that provides the described functionality: <https://atom.io/packages>

### Description

If you add a project folder which is not a git repo but it contains folders which are git repos, git diff won't work.
You can get it to work by adding the git repos as project folders, but it should be above the main project folder.

### Steps to Reproduce

1. Create the following tree:
|-- mainFolderWithoutGit
.....|-- someFolderWithGit
..........|-- .git
2. Add mainFolderWithoutGit as a project.
3. Make any changes to any file in someFolderWithGit

**Expected behavior:**

Changes should be detected by git-diff package

**Actual behavior:**

Changes are not detected

**Reproduces how often:**

100%

### Versions

1.52 and 1.53-beta0

### Additional Information

Fixed it here: https://github.com/atom/atom/pull/21631
Corresponding issue in tree-view: https://github.com/atom/tree-view/issues/1366

### PR DESCRIPTION
**Pull request by utkarshgupta137**
# Identify the Bug

Fixes https://github.com/atom/atom/issues/21630.
Fixes https://github.com/atom/github/issues/1149.
Fixes https://github.com/atom/github/issues/1835.

Linked PRs:
tree-view: https://github.com/atom/tree-view/pull/1367
tabs: https://github.com/atom/tabs/pull/560
status-bar: https://github.com/aminya/status-bar/pull/4

minimap-git-diff: https://github.com/atom-minimap/minimap-git-diff/pull/31
file-icons: https://github.com/file-icons/atom-fs/pull/9
linter: https://github.com/steelbrain/linter/pull/1715

### Description of the Change

By using atom.project.repositoryForPath function instead of matching the path with added directories, subfolders will get identified.
Made necessary changes to referenced places to handle the async nature of the atom.project.repositoryForPath.

### Alternate Designs

None.

### Possible Drawbacks

Since atom.project.repositoryForPath is an async function, there maybe a little lag when loading file which already has changes.

### Verification Process

TODO

### Release Notes

- Fixed subfolders of project folders not getting detected as git repositories.

### CODE REVIEW

**Comment by utkarshgupta137** [comment]
TODO: Add tests.

**Comment by sadick254** [comment]
@utkarshgupta137 Thanks for the contribution. It would really help If you could add tests on this.

**Comment by utkarshgupta137** [comment]
@sadick254 Done. I've also updated the corresponding PR in tabs package: https://github.com/atom/tabs/pull/560.
There is a corresponding PR in tree-view as well (https://github.com/atom/tree-view/pull/1367), but it Is taking me some time to get the tests running.

I'm not sure the 1 test failing for this PR is related to this PR at all.

**Review by M3TIOR — CHANGES_REQUESTED** [review]
Why are you modifying an existing Unit Test? This invalidates the purpose of the test you're modifying, and to make matters worse, the test you've turned it into couldn't be a conclusive test because the sample environment you're using is a direct copy of the initial container.

What should be done is as follows:
  1. Make a new copy of the existing test suite
  2. **Do as you've done here within that new copy of test suites** and add the child working directory.
  3. Grab a hook in each of the new tests pointing to the child repository so you can track that the child is being modified after every insertion and deletion, so on and so forth.
 
As an alternative you could have done steps 2 and 3 in every existing Unit Test, but that would have invalidated the Unit Test scheme. You should have a test for every feature. Since you're making a modification that coppies an existing set of features to a new target, you need to do the same to prevent regression.

File: packages/git-diff/spec/git-diff-spec.js
Comment (M3TIOR) [inline packages/git-diff/spec/git-diff-spec.js:120]: This is bad.

**Comment by M3TIOR** [comment]
Also sorry for being so awkward with the review, I'm really new to submitting reviews so the UI's a bit foreign at the moment.

File: packages/git-diff/spec/git-diff-spec.js
Comment (M3TIOR) [reply packages/git-diff/spec/git-diff-spec.js:120]: Damnit I highlighted more lines than that! Where's the container Unit Test in question Github!?

File: packages/git-diff/lib/helpers.js
Comment (aminya) [inline packages/git-diff/lib/helpers.js:3]: This doesn't need to be a heap defined lambda

```suggestion
async function repositoryForPath(path) {
```

**Review by aminya — APPROVED** [review]

**Comment by patricecoulon** [comment]
REMOVE KEBAB !
FUCK ISLAM !
ALLAH IS A PORC !

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by M3TIOR** [comment]
I don't care if it's joking. This is an environment intended for cooperative attitude and support. Not blatant racism. I'm reporting you @patricecoulon

**Comment by utkarshgupta137** [comment]
@M3TIOR I've modified the existing test case to see if git-diff works when the repository is inside another folder. If the git-diff is working on a repository inside a project folder, then it will surely work for any other repository in any other folder structure. I do understand that I've converted the test condition from necessary to sufficient for this case. If you want me to, I can separate the two tests cases/suites.

**Comment by M3TIOR** [comment]
I think that would be in the best interest of the project.

The following are some good principals to follow when writing unit tests IMO, but I deffer to the Atom maintainers since they have the final say anyhow:
 * No known and verified reproducible project issue, should ever be without a unit test to prevent regression once it's fixed. Unless it's something too complicated to test for like a memory leak.
 * No one should assume that a feature working in their environment will absolutely work in another until it's tested.
 * And finally, **unit** tests should only test individual features. It's in the name. 
 
Trying really hard to think of a way to word that last part that doesn't make me sound like a jerk. It's understandable how it could go over your head if English isn't your first language. But I can't really assume that it isn't your first language because most English speakers where I'm from don't really use it formally where I read it. And I have Aspergers so trying to parse informal English is a bit rough. What I'm trying to say, is that Islamophobic jerk who showed up here earlier, has made it impossible for me to feel as though I'm communicating in a way that's not insensitive.

**Comment by utkarshgupta137** [comment]
@M3TIOR I've duplicated the test suite with changes required for testing subfolder detection & have kept the original test suite as is.

**Review by M3TIOR — CHANGES_REQUESTED** [review]
So far, it doesn't look like there's any way to test what we need to in any reasonable manner. Unfortunately it will have to be hacked together. As much as I don't like that. Since the test suite needed is very complicated to administer, this will be good enough for now once you have modified the names of the tests to appropriately convey their unique purposes.

If you're willing to go the extra mile and make the test 100% regression-proof, I'm sure the Atom maintainers would appreciate it though.

**Comment by sadick254** [comment]
Hey @utkarshgupta137. Apologies for not looking into this on time. I know this PR has been here for a long time, during that time Atom has seen a number of changes. Due to the changes, this PR has now become outdated. Is it possible for you to resolve the conflicts?

**Comment by utkarshgupta137** [comment]
@sadick254 Done.

**Comment by sadick254** [comment]
@utkarshgupta137  Thank you for resolving the conflicts. This looks good, however, there are some comments from @M3TIOR  that have not been resolved. Can we get those resolved?

**Review by joMarot — CHANGES_REQUESTED** [review]
It's not all green.

**Comment by M3TIOR** [comment]
I rebased my spec for this commit. A brief note: I don''t know what happened between the last Atom version I worked on and the current development version though I could look through the commit history and trudge through to find a solution, but I really don't have the energy, please forgive my laziness at this juncture. For some reason all my async Jasmine specs are breaking. Specifically timing out repeatedly, and I can't get Jasmine to give me a longer timeout to see if there are any real bugs in my specs.

With that said, I hope this helps.

https://github.com/utkarshgupta137/atom/pull/1

**Comment by utkarshgupta137** [comment]
@sadick254 It looks like the tests are failing at an unrelated stage: https://github.visualstudio.com/Atom/_build/results?buildId=116370&view=logs&j=e4115def-01ab-568c-a31a-bc15166ffabd&t=c4607463-cf32-4eef-9650-44237e8b1b61&l=542. Otherwise, it is good to go.

**Comment by M3TIOR** [comment]
As much as I'd like to write off the errors within the module as a consequence of core modifications; why is it that only core tests are being run @sadick254 ? Shouldn't Jasmine be running the full sub-module tests? Maybe this comes down to a resource availability problem, or a test suite limitation, but if a submodule was edited in the last commit it should be tested as well as the core. Otherwise you may be introducing regressions.

I could potentially start working on a patch for the test suite to detect what the git repositories last modifications were, and automate the test suite to run their specs during feature patches for those. If you think that's a good idea anyway.

**Review by HoaBiNgan21 — APPROVED** [review]

**Comment by darangi** [comment]
> @sadick254 It looks like the tests are failing at an unrelated stage: https://github.visualstudio.com/Atom/_build/results?buildId=116370&view=logs&j=e4115def-01ab-568c-a31a-bc15166ffabd&t=c4607463-cf32-4eef-9650-44237e8b1b61&l=542. Otherwise, it is good to go.

@utkarshgupta137 I re-ran the tests on the pipeline and it looks like the changes from this PR is definitely causing the failure, would you mind taking a look?

**Review by Lugas1988 — APPROVED** [review]

**Comment by M3TIOR** [comment]
So, this may come down to something I did. Either when I rebased I did so over the wrong base; ie from a dev branch. Whatever the reason, I double-checked my original modifications on the release branches and they work. I'll work on rebasing this again with a stable branch like v1.59.0-dev or v1.58.0 some time in the near future.

**Review by pixlephonexl — APPROVED** [review]

**Review by pixlephonexl — APPROVED** [review]

**Comment by sadick254** [comment]
This PR has been superseded by #23446

**Pull request closed without merging**
